File: project-euler/581/euler_581_v2.py
#!/usr/bin/env python

# The Expat License
#
# Copyright (c) 2017, Shlomi Fish
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import sys
from pybst.avltree import AVLTree
if sys.version_info > (3,):
    long = int


def main():
    primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, ]
    t = AVLTree([(x, True) for x in primes])
    prev = 1
    s = 0
    while True:
        n = t.get_min().key
        t.delete(n)
        if (n == prev + 1):
            s += prev
            print("Found s = %d for n = %d" % (s, prev))
            sys.stdout.flush()
        for x in primes:
            # mul = x * n
            # if not t.get_node(mul):
            # t.insert(mul, True)
            t.insert(x * n, True)
        prev = n


main()

File: project-euler/581/euler_581_v2.cpp
#include <set>
#include <array>
#include <cstdio>
#include <set>

typedef long long ll;

int main()
{
    std::array<ll, 15> primes{ {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47} };
    std::set<ll> t;
    for (auto x: primes)
    {
        t.insert(x);
    }
    ll prev = 1;
    ll s = 0;
    while (true)
    {
        auto n = *(t.begin());
        t.erase(n);
        if (n == prev + 1)
        {
            s += prev;
            printf("Found s = %lld for n = %lld\n", s, prev);
            fflush(stdout);
        }
        for (auto x: primes)
        {
            t.insert(x * n);
        }
        prev = n;
    }

    return 0;
}
